README.adoc

```diff
@@ -8,42 +8,430 @@ The time for the full course is around three to four days.
+
+Advanced courses are tailored like base courses to the need of the
+group. For example, an basic course for for asynchronous programming
+explains how concurrent execution in Rust works, the advanced one may
+include writing your own execution engine.
+
+Topics
+------
+
+Ferrous Systems offers Rust courses for a wide variety of subjects. The
+following is a non-exhaustive list of subjects.
+
+Every subject with be taught with hands-on exercises.
+
+Core topics
+~~~~~~~~~~~
+
+This is the core component of the course. In this course, attendees will
+learn:
+
+* Ownership
+* Borrowing
+* Lifetimes
+* Working with memory and data structures
+* Control flow
+* Structuring of applications
+* Using and configuring cargo and rustup
+* Introducing Rust into existing products
+* General Rust patterns
+* Error Handling
+* Using Generic APIs
+* Using Rusts guarantees for security gains
+* Testing and Debugging
+* Documentation tooling
+* Basics of concurrency safety
+* Overview of important documentation
+* Overview the standard library, especially core interface like
+collections, input/output and networking
+* Overview of common libraries
+* ``Refactoring towards Speed'': a core technique to safely derive a
+fast program from a working program
+
+Exercises will be tailored to the chosen special subjects.
+
+Advanced Generic Programming in Rust
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Using libraries with generics in Rust is taught in the core course. When
+to use them, and how is a big question in library development, though.
+
+In this module, attendees will learn:
+
+* Introduction into advanced programming with generics in Rust
+* Writing generic APIs
+* Impact on compile time, size and runtime speed
+* Useful generics patterns
+* Patterns to avoid
+
+Rust in C/C++ Environments
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Rust is often deployed in existing products, especially withing or among
+solutions written in C/C++. This module explains binding efficiently
+from C/C++ to Rust and from Rust to C/C++ codebases.
+
+In this module, attendees will learn:
+
+* Rusts FFI (Foreign Function Interface)
+* `unsafe`, as needed for FFI
+* Safe binding, both manually and automatically
+* Binding strategies
+* Working with raw pointers and helping pointers
+* Costs of boundary crossing
+* Assessment of feasibility
+
+Unsafe Rust
+~~~~~~~~~~~
+
+Rusts `unsafe` feature is sometimes necessary for speed optimisations or
+implementation of special data structures. This module explains its
+position and use in the language.
+
+In this module, attendees will learn:
+
+* The role of `unsafe`
+* The scope of `unsafe`
+* Do’s and Don’ts of unsafe Rust
+* Introduction into support APIs, like non-null pointers
+* Potential undefined behaviour arising from the use of `unsafe`
+* Checking unsafe Rust vor safety
+
+Rust Testing
+~~~~~~~~~~~~
+
+This module teaches advanced Rust testing techniques.
+
+In this module, attendees will learn: * Fuzzing of Rust applications *
```

Comment:
Does this need to be broken into multiple lines?

Reply:
Yep, that's an error. Fixed.